### PR TITLE
Store evidences

### PR DESCRIPTION
**Pull request by mira13**
Store evidences Thibaut code adopted for otter

### CODE REVIEW

File: modules/Bio/Otter/ServerAction/Region.pm [hunk 1]
```diff
@@ -445,6 +451,522 @@ sub _write_region_exclusive { # runs under $slb->exclusive_work
...
+  my $pipeline_db = $self->server->dataset->pipeline_dba;
+
+
+  my $protein_adaptor = $pipeline_db->get_ProteinAlignFeatureAdaptor;
+  my $dna_adaptor = $pipeline_db->get_DnaAlignFeatureAdaptor;
+  my $transformed_exon = $exon->transform('seqlevel');
+
+  my $start_Exon;
+  my $end_Exon;
+  if ($transcript->translation) {
```
Comment (thibauthourlier) [inline modules/Bio/Otter/ServerAction/Region.pm]: move it out of cycle/function

File: modules/Bio/Otter/ServerAction/Region.pm [hunk 2]
```diff
@@ -355,9 +354,16 @@ sub _write_region_exclusive { # runs under $slb->exclusive_work
...
             $tran->slice($db_slice);
             $tran->transcript_author($author_obj);
+
+            foreach my $exon (@{ $tran->get_all_Exons }) {
+                $exon->slice($db_slice);
+                $self->fill_exon_align_evidence($exon, $tran);
+
+            }
         }
         foreach my $exon (@{ $gene->get_all_Exons }) {
```
Comment (thibauthourlier) [inline modules/Bio/Otter/ServerAction/Region.pm]: remove this cycle

File: modules/Bio/Otter/ServerAction/Region.pm [hunk 3]
```diff
@@ -445,6 +451,522 @@ sub _write_region_exclusive { # runs under $slb->exclusive_work
...
+          if ($toplevel_exon) {
+
+            $self->check_all_supporting_evidences($toplevel_exon, $target_slice, \%supporting_evidences, $exon, $start_Exon, $end_Exon, $pipeline_slice, $transcript);
+          }
+          else {
+            my $target_contig_projection;
+            eval {
+              $target_contig_projection = $target_slice_contig->project($target_slice->coord_system->name, $target_slice->coord_system->version);
+            };
+            if ($@) {
```
Comment (thibauthourlier) [inline modules/Bio/Otter/ServerAction/Region.pm]: remove this if block

File: modules/Bio/Otter/ServerAction/Region.pm [hunk 4]
```diff
@@ -445,6 +451,522 @@ sub _write_region_exclusive { # runs under $slb->exclusive_work
...
+            }
+            if ($target_contig_projection) {
```
Comment (thibauthourlier) [inline modules/Bio/Otter/ServerAction/Region.pm]: on else put error not warn

File: modules/Bio/Otter/ServerAction/Region.pm [hunk 5]
```diff
@@ -445,6 +451,522 @@ sub _write_region_exclusive { # runs under $slb->exclusive_work
...
+      $toplevel_exon->add_supporting_features(grep {defined $_} @$sfs);
+      foreach my $sf (values %$supporting_evidences) {
+#       We will try to project the contig to the top level, then for each of the top level region
+#       we will resize the feature to make it fit on the region and we will transfer the feature
+#       to the top level.
+        my $sf_slice = $sf->slice->sub_Slice($sf->start, $sf->end);
+        my $sf_on_sf_slice = $sf->transfer($sf_slice);
+        my $projected_slice = $sf_on_sf_slice->slice->project($target_slice->coord_system->name, $target_slice->coord_system->version);
+        if ($projected_slice) {
+          foreach my $elm (@$projected_slice) {
```
Comment (thibauthourlier) [inline modules/Bio/Otter/ServerAction/Region.pm]: remove empty loop

File: modules/Bio/Otter/ServerAction/Region.pm [hunk 6]
```diff
@@ -445,6 +451,522 @@ sub _write_region_exclusive { # runs under $slb->exclusive_work
...
+  }
+}
+
+
+sub check_all_supporting_evidences {
+  my ($self, $toplevel_exon, $target_slice, $supporting_evidences, $exon, $start_Exon, $end_Exon, $pipeline_slice, $transcript) = @_;
+# When we have transfered the exon, we need to make sure that it is on the top level sequence
+# we aim for and that the slice object is on the forward (1) strand. Otherwise it might cause
+# problems later
+  if ($toplevel_exon->slice->strand == -1) {
```
Comment (thibauthourlier) [inline modules/Bio/Otter/ServerAction/Region.pm]: refactor for just single if

File: modules/Bio/Otter/ServerAction/Region.pm [hunk 7]
```diff
@@ -445,6 +451,522 @@ sub _write_region_exclusive { # runs under $slb->exclusive_work
...
+              push(@sf_fragments, $sf_to_add);
+            }
+            else {
+              warn('WEIRD it was cut to fit');
+            }
+          }
+          foreach my $loutre_side_slice (@slices_to_project) {
+            my $loutre_side_projection = $loutre_side_slice->project('seqlevel');
+            if ($loutre_side_projection) {
+              foreach my $lsp (@$loutre_side_projection) {
```
Comment (thibauthourlier) [inline modules/Bio/Otter/ServerAction/Region.pm]: remove loop completly

File: modules/Bio/Otter/ServerAction/Region.pm [hunk 8]
```diff
@@ -445,6 +451,522 @@ sub _write_region_exclusive { # runs under $slb->exclusive_work
...
+                  if ($cut_sf->start < 1) {
+                    $cut_sf->start(1);
+                  }
+                  if ($cut_sf->end > $exon_projection_slice->length or $cut_sf->length > $exon_projection_slice->length) {
+                    $cut_sf->end($exon_projection_slice->length);
+                  }
+                  set_hstart_hend($sf, $cut_sf);
+                  $sf = $cut_sf;
+                }
+                else {
```
Comment (thibauthourlier) [inline modules/Bio/Otter/ServerAction/Region.pm]: remove empty

**Comment by mira13** [comment]
@thibauthourlier discussed changes applied, thanks

**Comment by s-mm** [comment]
I will review this once NoMerge has been deployed.

**Comment by s-mm** [comment]
@mira13 Have you made the changes that Thibaut suggested?

**Comment by mira13** [comment]
@s-mm yes, it's all done and agreed with Thibaut, see my prev comment

**Pull request merged**